### PR TITLE
Settings Save/Load System

### PR DESCRIPTION
**Pull request by SavanticIO**
Updated WCSharp.Json and WCSharp.Shared so that WCSharp.SaveLoad could be added.

Added a save and load system using WCSharp.SaveLoad for player settings.

Added commands for toggling Dialogue, QuestText and Captions

Added a command to see current settings

### CODE REVIEW

File: src/MacroTools/Commands/Captions.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+﻿using MacroTools.CommandSystem;
+using MacroTools.Extensions;
+using static War3Api.Common;
+
+namespace MacroTools.Commands
+{
+  /// <summary>
+  /// A <see cref="CommandSystem.Command"/> that sets the player's camera to a specific height.
```
Comment (YakaryBovine) [inline src/MacroTools/Commands/Captions.cs:8]: This comment refers to camera stuff.

File: src/MacroTools/Commands/Captions.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    /// <inheritdoc />
+    public override string Description => "Toggles the display of captions.";
+
+    /// <inheritdoc />
+    public override string Execute(player commandUser, params string[] parameters)
+    {
+      var captions = parameters[0];
+      if (!bool.TryParse(captions, out var captionsBool))
+        return "You must specify a boolean value(true or false) as the first parameter.";
```
Comment (YakaryBovine) [inline src/MacroTools/Commands/Captions.cs:32]:
```suggestion
        return "You must specify either true or false as the first parameter.";
```

File: src/MacroTools/Commands/Dialogue.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+﻿using MacroTools.CommandSystem;
+using MacroTools.Extensions;
+using static War3Api.Common;
+
+namespace MacroTools.Commands
+{
+  /// <summary>
+  /// A <see cref="CommandSystem.Command"/> that sets the player's camera to a specific height.
```
Comment (YakaryBovine) [inline src/MacroTools/Commands/Dialogue.cs:8]: Not the right comment.

File: src/MacroTools/Commands/Dialogue.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  /// </summary>
+  public sealed class Dialouge : Command
```
Comment (YakaryBovine) [inline src/MacroTools/Commands/Dialogue.cs:10]:
```suggestion
  public sealed class Dialogue : Command
```

File: src/MacroTools/Commands/Captions.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    public override bool Exact => false;
+  
+    /// <inheritdoc />
+    public override int MinimumParameterCount => 1;
+
+    /// <inheritdoc />
+    public override CommandType Type => CommandType.Normal;
+
+    /// <inheritdoc />
+    public override string Description => "Toggles the display of captions.";
```
Comment (YakaryBovine) [inline src/MacroTools/Commands/Captions.cs:25]: Let's keep the language really simple for people.
```suggestion
    public override string Description => "Turns dialogue captions on or off.";
```

File: src/MacroTools/Commands/Settings.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    public override bool Exact => false;
+  
+    /// <inheritdoc />
+    public override int MinimumParameterCount => 0;
+
+    /// <inheritdoc />
+    public override CommandType Type => CommandType.Normal;
+
+    /// <inheritdoc />
+    public override string Description => "Shows the current settings for the player.";
```
Comment (YakaryBovine) [inline src/MacroTools/Commands/Settings.cs:22]:
```suggestion
    public override string Description => "Shows your current settings.";
```

**Review by YakaryBovine — CHANGES_REQUESTED** [review]
Really, really cool stuff. Nothing to add in terms of the code I think. I just have a handful of suggestions for tooltips. Nice work!

**Comment by SavanticIO** [comment]
Thanks. I’ve tested the loading, saving and a failed load. I haven’t tested the failedhash response. I think you would have to correctly deserialise, change a value and then serialise again but not 100% sure on that.

**Review by YakaryBovine — APPROVED** [review]
Awesome. Happy with this!

**Pull request merged**